lib/open_food_network/sales_tax_report.rb

```diff
@@ -66,7 +66,7 @@ def table
 
     def relevant_rates
       return @relevant_rates unless @relevant_rates.nil?
-      @relevant_rates = Spree::TaxRate.pluck(:amount).uniq
+      @relevant_rates = Spree::TaxRate.uniq
```

Comment:
This is quite dangerous due to the query that generates:

```
irb(main):001:0> Spree::TaxRate.uniq.to_sql
=> "SELECT DISTINCT \"spree_tax_rates\".* FROM \"spree_tax_rates\" "
```

If you take a look at the explain:

```sql
openfoodnetwork=# EXPLAIN SELECT DISTINCT "spree_tax_rates".* FROM "spree_tax_rates";
                                                       QUERY PLAN
------------------------------------------------------------------------------------------------------------------------
 HashAggregate  (cost=14.23..15.53 rows=130 width=560)
   Group Key: id, amount, zone_id, tax_category_id, created_at, updated_at, included_in_price, name, show_rate_in_label
   ->  Seq Scan on spree_tax_rates  (cost=0.00..11.30 rows=130 width=560)
(3 rows)
```

It performs a sequential scan on the `spree_tax_rates` table, which can lead to 💥 (or serious performance hit). As I presume that table does not have too many rows I suggest we create an issue for it. The point is to limit the data we fetch from DB instead of blindly getting everything on that table

Reply:
Getting only the relevant tax rates was what I wanted to do in the first place, but it appeared that it was both costly and mostly useless.
Costly because the report gets the orders that are searched for and gets the corresponding tax rates and amounts via the previously discussed method (L47-50). To build the table we would need those orders/tax rates/amounts beforehand. But maybe we could memorize it (in `#search`) like we do in `#relevant_rates`?
By the way the use of `distinct` seems to lower downs the postgres "cost" and I guess it's useless as there wouldn't be any duplicate rows if there's no joint, right?
It also seemed mostly useless because the `spree_tax_rates` wouldn't be very large, typically there wouldn't be more 2 or 3 different taxes, don't you think? The sequential scan also seems pretty normal as we're doing a `SELECT *` so PG is browsing the whole table.
Happy to do what you think is best, I guess memorizing the search results to use it in `#header` and `#table` might be nice but only if Rails made the queries for `spree_tax_rates` only once for both times we would call `orders.tax_adjustment_totals` (or `orders.map(&:tax_adjustments).flatten.map(&:tax_rates).flatten.uniq` for what it's worth but that's not really pretty although we could create `Order#tax_rates` for this) ; but I think it would run the queries twice.

Reply:
I was just asking to create an issue so that we can work on it some other day. To answer your questions:

> To build the table we would need those orders/tax rates/amounts beforehand. But maybe we could memorize it

Exactly, and we can fetch the ones we need as we do with the orders in `#search`.

> By the way the use of distinct seems to lower downs the postgres "cost" and I guess it's useless as there wouldn't be any duplicate rows if there's no joint, right?

If there were no duplicates then there would be no need for the call to `#uniq` in line 69. However, that's not enforced at DB level and so it could happen. In general terms, `DISTINCT` cuts down the cost because fewer rows need to be fetched from disk and sent over the network.

> It also seemed mostly useless because the spree_tax_rates wouldn't be very large, typically there wouldn't be more 2 or 3 different taxes, don't you think?

That's what I guess but nothing stops anyone from creating lots of them. Besides, I'm not sure if that is something created by the instance admin and shared among enterprises or if it's something that each enterprise manager creates. In the latter case, it'd much easier to end up having lots of taxes on DB table.

> The sequential scan also seems pretty normal as we're doing a SELECT * so PG is browsing the whole table.

What makes PG choose a particular [Query Plan](http://www.postgresguide.com/performance/explain.html) are the constraints specified in the query's `WHERE` clause. If you specify a constraint on a column that has an index, chances are that it'll pick it up instead of a sequential scan. That is, the scan will happen on the index structure rather than on the table, which is way faster.

What `SELECT` can help with, when selecting a particular column/s, is reducing the cost of fetching the actual data from disk, the cost of sending it over the network and the number of Ruby objects instantiated in memory (which normally has a big impact on performance).

> I guess memorizing the search results to use it in #header and #table might be nice but only if Rails made the queries for spree_tax_rates only once for both times we would call orders.tax_adjustment_totals (or orders.map(&:tax_adjustments).flatten.map(&:tax_rates).flatten.uniq for what it's worth but that's not really pretty although we could create Order#tax_rates for this) ; but I think it would run the queries twice.

If the query stays the same for both `#header` and `#table` we can just save its result in a var. In any case, I would discourage doing `orders.map(&:tax_adjustments).flatten.map(&:tax_rates).flatten.uniq` performance wise. That's exactly 5 times going through `orders`, while we can do the same leveraging Postgres performance and the declarative nature of SQL. We would just ask for the data we need without loops. 

Also, things like these have become a big problem in reports. `5 x 10` orders it's not a big deal, but when fetching 1000 orders, for instance, performance starts to be affected.

Reply:
> I was just asking to create an issue so that we can work on it some other day.

Oh sorry I misunderstood.

> Besides, I'm not sure if that is something created by the instance admin and shared among enterprises or if it's something that each enterprise manager creates.

Yes they're created by the admin, and enterprise managers only choose which one applies.

> What makes PG choose a particular Query Plan are the constraints specified in the query's WHERE clause.

Yeah sorry I meant "SELECT without WHERE" ^.^

>  I would discourage doing orders.map(&:tax_adjustments).flatten.map(&:tax_rates).flatten.uniq performance wise. That's exactly 5 times going through orders, while we can do the same leveraging Postgres performance and the declarative nature of SQL.

Actually I just remembered that I had the same worry concerning fetching the tax rates and I wrote [something like what you say](https://github.com/openfoodfoundation/openfoodnetwork/blob/d22f5678bea35e8ef607f0ee7c3e30770190c695/lib/open_food_network/sales_tax_report.rb#L69-L71) (it made just one query and that was nice!).
... but I removed it afterwards because it didn't work with taxes applied on shipping fees and adjustments created from Spree::Admin::AdjustmentsController#create (see [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/controllers/spree/admin/adjustments_controller_decorator.rb#L31-L40)), because they don't "link" to the corresponding tax rate and only have the applied tax amount in their `included_tax` (that's why I had to write [this method](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/models/spree/adjustment_decorator.rb#L61-L68). I had no clue how to make this fit with the "relevant tax rates" query so I resigned to use `TaxRate.all` ><